### PR TITLE
feat: [PL-40084]: Add an api to disable public access on resource

### PR DESCRIPTION
**Pull request by meenakshi-harness**
## Describe your changes

Added API to disable public access on a given resource. ( We will extend this method to support scope deletion as well in followup iterations)

Testing:
1. Disable public access on resource.
2. If resourceId not provided throw error.

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/52726)

### CODE REVIEW

File: access-control/contracts/openapi/v1/openapi.yaml [hunk 1]
```diff
@@ -103,6 +103,47 @@ paths:
...
+        - $ref: '#/components/parameters/AccountHeader'
+        - schema:
+            type: string
+          in: query
+          name: account
+          description: Account Identifier
+        - schema:
+            type: string
+          in: query
+          name: org
```
Comment (mankrit-singh) [inline access-control/contracts/openapi/v1/openapi.yaml:127]: can these be path params? (org and project)

File: access-control/contracts/openapi/v1/openapi.yaml [hunk 2]
```diff
@@ -103,6 +103,47 @@ paths:
...
+          $ref: '#/components/responses/PublicAccessResponse'
+      description: Disable Public Access
+      security:
+        - x-api-key: []
+      parameters:
+        - $ref: '#/components/parameters/AccountHeader'
+        - schema:
+            type: string
+          in: query
+          name: account
```
Comment (mankrit-singh) [inline access-control/contracts/openapi/v1/openapi.yaml:122]: do we need this? It's already present in Account Header and x-api-key

File: access-control/contracts/openapi/v1/openapi.yaml [hunk 3]
```diff
@@ -103,6 +103,47 @@ paths:
       tags:
         - Public Access
       x-internal: true
+  /v1/public-access/disable:
+    parameters: []
+    put:
+      summary: Update a Role
```
Comment (mankrit-singh) [inline access-control/contracts/openapi/v1/openapi.yaml:109]: This should not be Update a Role or PUT call

File: access-control/contracts/openapi/v1/openapi.yaml (hunk 1 above)
Comment (meenakshi-harness) [reply access-control/contracts/openapi/v1/openapi.yaml:127]: No we need them as query params only.

File: access-control/contracts/openapi/v1/openapi.yaml (hunk 2 above)
Comment (meenakshi-harness) [reply access-control/contracts/openapi/v1/openapi.yaml:122]: Yes account here is to specify which scope we want to disable public access. Have validation check in BE for both.

File: access-control/contracts/openapi/v1/openapi.yaml (hunk 3 above)
Comment (meenakshi-harness) [reply access-control/contracts/openapi/v1/openapi.yaml:109]: Sure will update summary. Discussed with @ashishsanodia-harness to keep it as PUT.

**Comment by meenakshi-harness** [comment]
trigger uts

**Review by mankrit-singh — APPROVED** [review]

**Comment by piyush-harness** [comment]
*[`access-control/service/src/main/java/io/harness/accesscontrol/publicaccess/PublicAccessServiceImpl.java` line 146 at r2](https://reviewable.io/reviews/harness/harness-core/52726#-Ne9Akq4CSt_1uBZ4qR7:-Ne9Akq4CSt_1uBZ4qR8:b-7nesjx) ([raw file](https://github.com/harness/harness-core/blob/69d501dc9c4d235825fcfc396817cc072d7ce19b/access-control/service/src/main/java/io/harness/accesscontrol/publicaccess/PublicAccessServiceImpl.java#L146)):*
> ```Java
>     scope.setAccount(account);
>     scope.setProject(project);
>     scope.org(org);
> ```

Why not setOrg?

**Review by Jimit09Gandhi — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52726)* status: 5 file(s) to review. 1 change requests. 3. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`access-control/service/src/test/java/io/harness/accesscontrol/publicaccess/PublicAccessServiceImplTest.java` line 249 at r2](https://reviewable.io/reviews/harness/harness-core/52726#-Ne9DDSi9zWgbRGZJVdQ:-Ne9DDSi9zWgbRGZJVdR:bpez2g5) ([raw file](https://github.com/harness/harness-core/blob/69d501dc9c4d235825fcfc396817cc072d7ce19b/access-control/service/src/test/java/io/harness/accesscontrol/publicaccess/PublicAccessServiceImplTest.java#L249)):*
> ```Java
>     when(NGRestUtils.getResponse(request)).thenReturn(resourceGroupResponse);
> 
>     boolean result = publicAccessService.disablePublicAccess(
> ```

assert on result.

**Review by meenakshi-harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52726)* status: 5 file(s) to review. 1 change requests. 5. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`access-control/service/src/main/java/io/harness/accesscontrol/publicaccess/PublicAccessServiceImpl.java` line 146 at r2](https://reviewable.io/reviews/harness/harness-core/52726#-Ne9Akq4CSt_1uBZ4qR7:-Ne9H55t6_HmOultJLer:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/69d501dc9c4d235825fcfc396817cc072d7ce19b/access-control/service/src/main/java/io/harness/accesscontrol/publicaccess/PublicAccessServiceImpl.java#L146)):*
<details><summary><i>Previously, piyush-harness (Piyush Patel) wrote…</i></summary><blockquote>

Why not setOrg?
</blockquote></details>

Done.
___
*[`access-control/service/src/test/java/io/harness/accesscontrol/publicaccess/PublicAccessServiceImplTest.java` line 249 at r2](https://reviewable.io/reviews/harness/harness-core/52726#-Ne9DDSi9zWgbRGZJVdQ:-Ne9H60WAK_AKt69cFtm:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/69d501dc9c4d235825fcfc396817cc072d7ce19b/access-control/service/src/test/java/io/harness/accesscontrol/publicaccess/PublicAccessServiceImplTest.java#L249)):*
<details><summary><i>Previously, Jimit09Gandhi (Jimit Gandhi) wrote…</i></summary><blockquote>

assert on result.
</blockquote></details>

Done.

**Review by Jimit09Gandhi — APPROVED** [review]
> Reviewed 1 of 4 files at r1, 2 of 4 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52726)* status: 5 file(s) to review. 4. unresolved discussions. 1 of 2 approvals obtained.

**Review by ashishsanodia-harness — APPROVED** [review]
:lgtm:

> Reviewed 1 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52726#-:-Ne9lC_H5LEnRKV38-2T:bnfp4nl)* status: 4 file(s) to review. 4. unresolved discussions.

**Pull request merged**